Use quality header and SAMTag.CO in FastqEncoder

magicDGS opened the pull request:
### Description

Please explain the changes you made here.
Explain the **motivation** for making this change. What existing problem does the pull request solve?

### Checklist

- [x] Code compiles correctly
- [x] New tests covering changes and new functionality
- [x] All tests passing
- [ ] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

lbergelson commented:
@magicDGS Could you explain this one?  It makes sense to capture the comment lines from fastq somewhere, but is CO the standard place?  

@yfarjoun Any objections to this?  It could potentially increase bam size if we have comments in the fastq, but A) I assume we don't, B) I don't think we even go through fastq anymore?

magicDGS commented:
@lbergelson - specifications have the `CO` tag defined as a [free-text comment](https://samtools.github.io/hts-specs/SAMtags.pdf). Some FASTQ files that I found in the wild include in the quality header (starting with `+`) comments about the read. I thought that these comments can leave on the `CO` tag as it is defined as comments. Other information that can be stored in the `CO` tag are the comments in the read name, but that's more tricky as some of this information might be barcodes or pair-end information.

Another option is to design a new tag in the specs for the FASTQ quality-header (`FQ`?), and use the `CO` for other comments, such the ones in the read name (after the space). What do you think that is better, @lbergelson/@yfarjoun? I would like to have a lossless way of transforming FASTQ to SAM, and store this kind of information in tags is the best way...

yfarjoun commented:
To be more accurate, the header CO tag is specified as 
> One-line text comment. Unordered multiple @CO lines are allowed. UTF-8 encoding may be used.

the "free text" is from the alignment tag "CO"...

magicDGS commented:
The suggestion is for a per-alignment `CO` tag, containing the information from the "quality header" from the FASTQ. This could vary between reads.

yfarjoun commented:
Ah, I see. sorry, my bad.

yfarjoun commented:
I'd prefer if you made the conversion of quality header to comment optional (default false).

magicDGS commented:
Sounds reasonable - done in next commit.

yfarjoun commented on src/main/java/htsjdk/samtools/fastq/FastqEncoder.java:
```diff
@@ -91,13 +92,25 @@ public static FastqRecord asFastqRecord(final SAMRecord record) {
             readName += (record.getFirstOfPairFlag()) ? FastqConstants.FIRST_OF_PAIR : FastqConstants.SECOND_OF_PAIR;
         }
-        return new FastqRecord(readName, record.getReadString(), null, record.getBaseQualityString());
+        return new FastqRecord(readName, record.getReadString(), record.getStringAttribute(SAMTag.CO.name()), record.getBaseQualityString());
```
Are you concerned that a record that has a comment that it not related to the QualityHeader will be misunderstood and the comment will then become the quality header? 

The wikipedia page seems to imply that the sample_id needs to be the "first" part of this header, followed by arbitrary text, but on the other hand, the fastq format is poorly defined...so it's not even clear what the separator is.

magicDGS replied:
I just would like to have a lossless conversion from FASTQ to SAM with this encoder. Validation of the comment could be performed downstream.

yfarjoun commented on src/main/java/htsjdk/samtools/fastq/FastqEncoder.java:
```diff
@@ -106,6 +119,9 @@ public static SAMRecord asSAMRecord(final FastqRecord record, final SAMFileHeade
         samRecord.setReadName(readName);
         samRecord.setReadBases(record.getReadBases());
+        if (qualityHeaderToComment) {
+            samRecord.setAttribute(SAMTag.CO.name(), record.getBaseQualityHeader());
```
make sure that tabs are not part of this string....as you'll destroy the bam otherwise. 
(please test this)

magicDGS replied:
Actually, this is a difficult issue: in bwa, for example, the quality-header could be use to assign tags to SAM/BAM. From http://bio-bwa.sourceforge.net/bwa.shtml:

```
-C | Append append FASTA/Q comment to SAM output. This option can be used to transfer read meta information (e.g. barcode) to the SAM output. Note that the FASTA/Q comment (the string after a space in the header line) must conform the SAM spec (e.g. BC:Z:CGTAC). Malformated comments lead to incorrect SAM output.
```

Maybe the way to go here is to parse the quality header for tag-formatted comments and if present add to the attributes; otherwise, add to the `SAMTag.CO` by converting any tab to spaces. What do you think?

yfarjoun replied:
I think that if there's some standard, (e.g. the one in bwa) we should use it. I don't like the idea of "either/or"....can you just use SAM-style tags and not allow for free text?

magicDGS replied:
I don't think that the `bwa` quality header is a standard, but an ad-hoc solution to allow including per-read information from FASTQ input reads into the output of `bwa`. I still would like to allow for free-text, because in the wild there are FASTQ files with that kind of header, and there are options when downloading from SRA/ENA servers to include free-text information in quality headers (e.g., sratoolkit's `fastq-dump` command allows custom `--defline-qual`, which won't complies to SAM-style tags).

yfarjoun replied:
In that case, I would advise for a parameter switch for the different modes, rather than something that depends on "if present".....

magicDGS replied:
I added a even better configuration mode: use a `BiConsumer` to add extra information from the `FastqRecord` to the `SAMRecord` is a custom way: I included an implementation for the `CO` tag (converting tabs into spaces) and we can also include the one for setting the tags if you feel that is necessary.

magicDGS replied:
I did add also another custom encoder.

yfarjoun commented on src/test/java/htsjdk/samtools/fastq/FastqEncoderTest.java:
```diff
@@ -72,7 +72,18 @@ public void testAsSAMRecord() throws Exception {
+
+    @Test
+    public void testAsSAMRecordParsedTags() throws Exception {
+        final String tags = "BC:Z:ACTG\tRG:Z:sample1";
```
please test other types as well..

magicDGS replied:
Done (concretely integer and character)

yfarjoun commented on src/test/java/htsjdk/samtools/fastq/FastqEncoderTest.java:
```diff
@@ -46,13 +46,15 @@ public void testAsFastqRecord() throws Exception {
         testRecord(record.getReadName() + FastqConstants.SECOND_OF_PAIR, FastqEncoder.asFastqRecord(record), record);
         record.setSecondOfPairFlag(false);
         testRecord(record.getReadName(), FastqEncoder.asFastqRecord(record), record);
+        record.setAttribute("CO", "Comment in SAM tag");
```
use `SAMTag.CO.name()`

magicDGS replied:
Done

yfarjoun commented on src/test/java/htsjdk/samtools/fastq/FastqEncoderTest.java:
```diff
@@ -46,13 +46,15 @@ public void testAsFastqRecord() throws Exception {
         Assert.assertEquals(fastqRecord.getBaseQualities(), samRecord.getBaseQualities());
         Assert.assertEquals(fastqRecord.getReadBases(), samRecord.getReadBases());
-        Assert.assertNull(fastqRecord.getBaseQualityHeader());
+        Assert.assertEquals(fastqRecord.getBaseQualityHeader(), samRecord.getStringAttribute("CO"));
```
`SAMTag.CO.name()`

magicDGS replied:
Done

yfarjoun commented on src/test/java/htsjdk/samtools/fastq/FastqEncoderTest.java:
```diff
@@ -65,12 +67,30 @@ public void testAsSAMRecord() throws Exception {
+        // default method does not include the comment header
+        testConvertedSAMRecord(FastqEncoder.asSAMRecord(fastqRecord, samRecord.getHeader()), samRecord);
+        // test with qualityHeaderToComment=true populates the CO tag
+        samRecord.setAttribute("CO", fastqRecord.getBaseQualityHeader());
```
`SAMTag.CO.name()`

magicDGS replied:
Done

yfarjoun commented on src/test/java/htsjdk/samtools/fastq/FastqEncoderTest.java:
```diff
@@ -65,12 +67,30 @@ public void testAsSAMRecord() throws Exception {
         Assert.assertEquals(converted.getReadName(), original.getReadName());
         Assert.assertEquals(converted.getBaseQualities(), original.getBaseQualities());
         Assert.assertEquals(converted.getReadBases(), original.getReadBases());
+        Assert.assertEquals(converted.getStringAttribute("CO"), original.getStringAttribute("CO"));
```
`SAMTag.CO.name()`

magicDGS replied:
Done

yfarjoun commented on src/test/java/htsjdk/samtools/fastq/FastqEncoderTest.java:
```diff
@@ -65,12 +67,30 @@ public void testAsSAMRecord() throws Exception {
+        final FastqRecord record = new FastqRecord(samRecord.getReadName(), samRecord.getReadBases(), tags, samRecord.getBaseQualities());
+        final SAMRecord converted = FastqEncoder.asSAMRecord(record, samRecord.getHeader(), FastqEncoder.QUALITY_HEADER_PARSE_SAM_TAGS);
+        testConvertedSAMRecord(converted, samRecord);
+        Assert.assertEquals(converted.getAttribute("BC"), "ACTG");
```
use SAMTag variable here as well. if they do not exist, then please add them.

magicDGS replied:
Done

magicDGS commented:
@yfarjoun - addressed comments and hopefully ready to go!
@lbergelson - now the default behaviour is the same, so no worries about increased BAM size

magicDGS commented:
Friendly ping here @yfarjoun

lbergelson commented:
@magicDGS Rerunning the tests since they failed with what looks like a transient issue.  I'll merge when they pass.  Thank you.

magicDGS commented:
Thanks!
